### PR TITLE
Make go.thethings.network/lorawan-stack/v3/cmd a module

### PR DESCRIPTION
**Pull request by johanstokking**
#### Summary

Replaces #2493 

Make `go.thethings.network/lorawan-stack/v3/cmd` a module

#### Changes

- Move `tools.go` into new `tools` mod
- Make `cmd` a module
- Build development tooling to `.bin` instead of `go run` which would use the repo's module

#### Notes for Reviewers

1. [ ] Check need for the replacements in `/go.mod` and `/cmd/go.mod`
2. [x] Make sure that we can use the tools in `/cmd/tools.go`. I'm also open to a `/tools` module
3. [x] Check if we need to merge some changes from github.com/robertkrimen/otto in github.com/TheThingsIndustries/otto 
4. [ ] Figure out how we can update `.bin` binaries if modules update in `tools/go.mod`; maybe always install on `make init`?

#### Checklist

- [ ] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by johanstokking** [comment]
Blocked by #2496

**Comment by johanstokking** [comment]
In order to use versioned tools, I would suggest;

- put `tools.go` in `cmd`, or move to `tools/main.go` and init new module
- install the tools via `go install` to a shared repo-local bin folder
- use the binaries from the bin folder instead of the `go run` construct in mage (via `execGo`)

Example Makefile;

```Makefile
GO = go
GOBIN = $(PWD)/.bin
export GOBIN

.PHONY: deps.dev
deps.dev: .venv
	@pushd cmd && cat tools.go | grep _ | awk -F'"' '{print $$2}' | xargs -tI % $(GO) install % && popd
```

And then `./bin/unconvert` etc

File: .mage/mage.make [hunk 1]
```diff
@@ -12,11 +12,16 @@
...
 
+MAKE ?= make
 MAGE ?= ./mage
+GOBIN = $(PWD)/.bin
+export GOBIN
 
-$(MAGE): magefile.go $(wildcard .mage/*.go)
-	GO111MODULE=on go install github.com/magefile/mage
-	GO111MODULE=on go run github.com/magefile/mage -compile $(MAGE)
+$(GOBIN):
```
Comment (johanstokking) [inline .mage/mage.make:20]: Should we turn this into `$(GOBIN): tools/go.mod` for it to track that file for changes to the tools and versions?

File: tools/main.go [hunk 2]
```diff
@@ -14,7 +14,7 @@
 
 // +build tools
 
-package tools
+package main
```
Comment (rvolosatovs) [inline tools/main.go:17]: Why?

**Comment by rvolosatovs** [comment]
The benefit of using `go run` to run commands is the automatic installation/versioning with no `Makefile` etc. maintenance cost. What's the benefit of using compiled binaries? How do you propose to keep them up-to-date?

**Comment by johanstokking** [comment]
> The benefit of using `go run` to run commands is the automatic installation/versioning with no `Makefile` etc. maintenance cost.

Yes, that definitely has benefits. But I do want to keep those modules (and replacements) out of the project root's `go.mod`. Two possible steps forward:

1. `go run` from the `tools` dir and account for that in the paths
2. Watch `tools/go.mod` in the Makefile so that changes trigger reinstallation

**Comment by rvolosatovs** [comment]
I'd vote for `1`. We could move all the mage stuff there as well, in fact, I believe. That way we'd have everything tooling-related with all the dependency hell isolated in `./tools` and a clean top-level module at root. I believe we don't need a `./cmd` module either this way.

**Comment by johanstokking** [comment]
OK, please try if that works.

**Comment by johanstokking** [comment]
> That way we'd have everything tooling-related with all the dependency hell isolated in `./tools` and a clean top-level module at root

Looking at replacements we have in this changeset;

- `gopkg.in/DATA-DOG/go-sqlmock.v1`: that's because of `github.com/heptiolabs/healthcheck` required by `pkg/component`
- `github.com/labstack/echo/v4`: we're on an old version because of https://github.com/TheThingsNetwork/lorawan-stack/issues/977. With #2382 we're closer to getting rid of Echo. @htdvisser can you file the issues to migrate away completely, as announced in that PR?
- `github.com/Azure/go-autorest` is quite challenging, see https://github.com/google/go-cloud/issues/1990 (this is our dependency) and https://github.com/kubernetes/client-go/issues/628
- `github.com/robertkrimen/otto` replaced by our fork: I keep pinging these guys (https://github.com/robertkrimen/otto/pull/283) but no luck there
- `github.com/grpc-ecosystem/grpc-gateway`: is this tooling?

**Pull request closed without merging**